citest/base/bindings.py

```diff
@@ -0,0 +1,376 @@
+      parser.add_argument(arg[0], **arg[1])
+    options = parser.parse_known_args()
+
+    config_parser = ConfigParser.RawConfigParser(defaults=self.__defaults)
+    config_parser.read(self.__config_files)
+    flags = {}
+    flags.update(vars(options[0]))
+    extra = options[1]
+    if infer and extra:
+      flags.update(self.__infer_options(config_parser, extra))
+    flags.update({key.lower(): value
+                  for key, value in self.__overrides.items()})
+
+    bindings = ConfigurationBindings(
+        config_parser, overrides=flags, section=section)
+    return bindings
+
+  def _exists(self, path):
+    """Helper function to determine if a path exists to facilitate testing."""
+    return os.path.exists(path)
+
+  def __add_configs_for_name(self, name, location, dirs, accumulator):
+    """Find zero or more config files for the given entity name.
+
+    Args:
+      name: [string] The basename of the entity to locate config for.
+      location: [string] The path where the entity is located.
+      dirs: [string] Other standard paths to look in.
+      accumulator: [list] The files found with highest precedence first.
+    """
+    config_name = name + '.config'
+    all_dirs = []
+    if location:
+      all_dirs.append(os.path.join(location, config_name))
+    all_dirs.extend([os.path.join(elem, config_name) for elem in dirs])
+    for path in all_dirs:
+      if self._exists(path):
+        accumulator.append(path)
+
+  def add_configs_for_class(self, klass):
+    """Add all the config files that the given class may use.
+
+    This is the closure of its subclasses and modules.
+    """
+    name = klass.__name__
+    if name in self.__visited_for_config:
+      return
+    self.__visited_for_config.add(name)
+
+    found_list = []
+    location = os.path.dirname(inspect.getfile(klass))
+    dirs = ['.', os.path.join(os.path.expanduser('~'), '.citest')]
+    self.__add_configs_for_name(name, location, dirs, found_list)
+
+    package_parts = klass.__module__.split('.')
+    module = package_parts[-1]
+    self.__add_configs_for_name(module, location, dirs, found_list)
+
+    package_paths = []
+    package = '.'.join(package_parts[:-1]) if len(package_parts) > 1 else None
+
+    while package and not package in self.__visited_for_config:
+      parts = package.split('.')
+      package_path = os.path.join(location, '__package__.config')
+      if not package_path in self.__visited_for_config:
+        self.__visited_for_config.add(package_path)
+        if os.path.exists(package_path):
+          package_paths.append(package_path)
+
+      self.__add_configs_for_name(parts[0], location, [], found_list)
+      self.__add_configs_for_name(package, None, dirs, found_list)
+      package = '.'.join(parts[:-1]) if len(parts) > 1 else None
+      location = os.path.dirname(location)
+    found_list.extend(package_paths)
+
+    # Add in reverse order so highest precedence is last
+    self.__config_files.extend(found_list[::-1])
```

Comment:
Clarify what this means for the config files in the directory structure. Further down in the tree means lower priority?